Ignore task items that are empty

digitalmoksha opened the pull request:
GFM does note recognize empty tasks as valid task items.  For example:

- [ ]
- [ ] one
- [ ] two

However when using `task_list`, clicking on the `one` task would try to change the empty task.

I've updated the regex to ignore empty task items.

digitalmoksha commented:
@deckar01 I ran into difficult running the JS tests, they didn't seem to run (or it's just my total lack of experience using `phantomjs`.  I added a test, but not sure it passes.

digitalmoksha commented:
fyi, we're seeing this problem on GitLab, and is detailed in issue https://gitlab.com/gitlab-org/gitlab/issues/30182

deckar01 commented:
The tests on master seem to be working locally. Your tests also pass locally although the build is failing due to a [linting error](https://travis-ci.org/deckar01/task_list/jobs/609007059#L883). Cherry picking just your tests onto master also passes. I introduced an error into the code and confirmed it caused the tests to fail as a sanity check :wink:.

This seems to indicate that either the regression test is written incorrectly, or that the gitlab issue it not being caused by these regexes.

digitalmoksha commented:
Hmm, if you don’t mind, try pulling over the change I made to `test_task_lists_behavior.html`. With the original JS regex it will fail (clicking on the first checkbox tries to change the first line, which is not a checkbox). The new regex seems to fix that. I agree the test is not the best. 

I changed the ruby regex mostly for parity with the JS version.

There may be another way to fix it. But I do think there is a parsing problem in JS that causes an empty task item to be considered a task item.

deckar01 commented:
The issue is caused by the markdown filter that runs before the task list filter. It trims whitespace off of the list item content. `TaskList.updateSource()` operates directly on the raw markdown, so the whitespace is still present.

The minimal change would probably be to change `line.match(@itemPattern)` to `line.trim().match(@itemPattern)` to mimic this behavior.

A better change would be to use a client side markdown parser first like the backend does. Parsing around code fences and escape sequences with regexes adds a lot of complexity and has been error prone.

digitalmoksha commented:
I removed the regex changes and added the `.trim()`, that does seem to fix it.  `test_task_lists_behavior.html` now works correctly.

I didn't add any tests though.  I am unable to get `phantomjs` to execute any test at all.  It either returns immediately (when using node 7.10.1) or it just hangs (when using node 12.13.0 and the latest phantomjs, 2.1.1).  I'm running on macOS 14

`phantomjs` is [now archived and no longer under development](https://github.com/ariya/phantomjs/issues/15344) since Mar 2018.

I agree, using a real commonmark parser would make it much more robust 🤔

deckar01 commented:
I opened #28 to investigate the phantom tests. It is probably overdue for a real test framework. Thanks!

deckar01 commented:
@digitalmoksha This fix has been released in version 2.2.1.

digitalmoksha commented:
Thanks @deckar01!
